zephyr/drivers/can/can_sccan.c

```diff
@@ -404,6 +415,74 @@ static void sc_can_rx_isr(const struct device *dev)
 	return;
 }
 
+static void sc_can_get_error_count(const struct sc_can_cfg *config,
+			      struct can_bus_err_cnt *err_cnt)
+{
+	uint32_t errcnt_reg;
+
+	errcnt_reg = sc_can_read32(config, SCCAN_ECNTR_OFFSET);
+	err_cnt->tx_err_cnt = SCCAN_TXECNT(errcnt_reg);
+	err_cnt->rx_err_cnt = SCCAN_RXECNT(errcnt_reg);
+}
+
+static int sc_can_get_state(const struct device *dev, enum can_state *state,
+			      struct can_bus_err_cnt *err_cnt)
+{
+	const struct sc_can_cfg *config = dev->config;
+	uint32_t status_reg;
+
+	status_reg = sc_can_read32(config, SCCAN_STSR_OFFSET);
+	switch (SCCAN_ESTS(status_reg)) {
+		case SCCAN_ESTS_CAN_DISABLE:
+			*state = CAN_STATE_STOPPED;
+			break;
+		case SCCAN_ESTS_ERROR_ACTIVE:
+			if (status_reg & SCCAN_EWRN) {
+				*state = CAN_STATE_ERROR_WARNING;
+			} else {
+				*state = CAN_STATE_ERROR_ACTIVE;
+			}
```

Comment:
`SCCAN_ESTS_ERROR_ACTIVE` is defined as  `0b01`.   `SCCAN_EWRN` is defined as `BIT(1)`.
This means that `status_reg` never have bit 1, no?

Reply:
No, `SCCAN_ESTS_ERROR_ACTIVE` is assigned to BIT 2 and BIT3.
However this register spec is too complicated.
Will improve later.

Reply:
I don't get it.  You have `#define SCCAN_ESTS_ERROR_ACTIVE  (1U)`.  What do you mean by "assigned to BIT 2 and BIT3"?

Reply:
Sorry for short explain.

The specifications of the CAN Status register are described below.
https://github.com/spacecubics/scobc-fpga-technical-reference-manual/blob/main/can_controller.org

`CAN Status Register (Offset: 0x0018)`

Bit1:  CAN_EWRN
BIT2-3: CAN_ESTS

In switch case condition, check the CAN_ESTS (Error Status by BIT2-3), but in `SCCAN_ESTS_ERROR_ACTIVE` case,
check the `CAN_EWRN` (Error Warning).

So we check the another bit.